sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubPartitionHashJoin.scala

```diff
@@ -0,0 +1,890 @@
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+
+import org.apache.spark.sql.catalyst.plans.{ExistenceJoin, FullOuter, InnerLike, JoinType}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * Iterator that can tell if the data size of all the batches from current position
+ * is larger than the specified `targetBatchSize`, and whether the batches are
+ * serialized.
+ */
+class BatchTypeSizeAwareIterator(
+    inputIter: Iterator[ColumnarBatch],
+    targetBatchSize: Long,
+    intputBatchSize: GpuMetric) extends Iterator[ColumnarBatch]
+  with Arm with TaskAutoCloseableResource {
+
+  assert(targetBatchSize >= 0,
+    s"Target batch size should not be negative, but get $targetBatchSize")
+
+  private val readBatchesQueue = ArrayBuffer.empty[ColumnarBatch]
+
+  // Initialize it early by prefetching one batch to avoid duplicate checks on each batch
+  private val anyBatchSerialized = {
+    if (inputIter.hasNext) {
+      readBatchesQueue += inputIter.next()
+    }
+    readBatchesQueue.nonEmpty &&
+      GpuShuffledHashJoinExec.isBatchSerialized(readBatchesQueue.head)
+  }
+
+  private lazy val getBatchSize: ColumnarBatch => Long = if (anyBatchSerialized) {
+    // Need to take care of this case because of the optimization introduced by
+    // https://github.com/NVIDIA/spark-rapids/pull/4588.
+    // Roughly return the serialized data length as the batch size.
+    SerializedTableColumn.getMemoryUsed
+  } else {
+    GpuColumnVector.getTotalDeviceMemoryUsed
+  }
+
+  override def close(): Unit = if (!closed) {
+    readBatchesQueue.safeClose()
+    readBatchesQueue.clear()
+    super.close()
+  }
+
+  override def hasNext: Boolean = readBatchesQueue.nonEmpty || inputIter.hasNext
+
+  override def next(): ColumnarBatch = {
+    if (!hasNext) throw new NoSuchElementException()
+    closeOnExcept {
+      if (readBatchesQueue.nonEmpty) {
+        readBatchesQueue.remove(0)
+      } else {
+        inputIter.next()
+      }
+    } { batch =>
+      intputBatchSize += getBatchSize(batch)
+      batch
+    }
+  }
+
+  /**
+   * Whether the data size of all the batches from current position is larger than
+   * the given `targetBatchSize`.
+   */
+  def isBatchesSizeOverflow: Boolean = {
+    var readBatchesSize = readBatchesQueue.map(getBatchSize).sum
+    while (inputIter.hasNext && readBatchesSize <= targetBatchSize) {
+      val batch = inputIter.next()
+      readBatchesQueue += batch
+      readBatchesSize += getBatchSize(batch)
+    }
+    readBatchesSize > targetBatchSize
+  }
+
+  /**
+   * Whether the batches in the input iterator are serialized.
+   */
+  def areBatchesSerialized: Boolean = anyBatchSerialized
+}
+
+object GpuSubPartitionHashJoin extends Arm {
+  /**
+   * The seed for sub partitioner for hash join.
+   * Differ from the default value: 0.
+   */
+  val SUB_PARTITION_HASH_SEED: Int = 100
```

Comment:
Could you share more about this ?
 
> What happens if the input data is still too large?

Does this mean the data sizes in one or some partitions after the first re-partitioning are still too large ?

Reply:
I am fine with doing the reparation only once to start out with, but long term we need to be able to re-partition multiple times and also have special cases to deal with skewed data. These should be follow on work because what you have now is much better than what we had before and I don't want to wait to get this in.

There are two situations that I am concerned about.

1. The build side batch is really-really-really big.  Currently we are re-partitioning the data 16 ways by default. If the build side is very large, like 64 GiB, then each build said batch would be about 4 GiB in size. This is larger than our target size so it would likely result in us using too much GPU memory to do the computation.  The idea would be that if we do hit a situation where one or more the build side batches are still too large we would try to re-partition them again, but with a different seed. We also would have seen all of the data so we could guess at a better number of partitions to try the second time. (This is really minor because the number of partitions is configurable, but I would like prefer to have us to not crash even in bad situations). 
2. There is skew in the join keys to the point that the build side is too large because there is only one key. This would be something that we would have to add as a part of possibly re-partitioning the data multiple times. Without it we could run into a live-lock situation where we keep trying to re-partition the data more, but nothing is happening when we do it. In this case in the short to medium term we would probably just try to do the join, and if it didn't work we would fail.  But long term we would want to try and implement a cross join what could do the join correctly, but might take longer than the optimized joins would.

Reply:
Will file an issue to track it as a follow-up.

Reply:
Track them by https://github.com/NVIDIA/spark-rapids/issues/7832. 
Free to update this issue if I miss anything.